File: hora/horoscope/dhasa/sudharsana_chakra.py
from hora.horoscope.chart import charts
from hora import const, utils
from hora.panchanga import panchanga
def sudharsana_chakra_dhasa_for_divisional_chart(jd_at_dob,place,dob,years_from_dob=0,divisional_chart_factor=1):
    """
        calculate sudharsana chakra dhasa for divisional charts / annual charts
        for just divisional charts - use divisional_chart_factor and set years_from_dob = 0
        for annual charts use years_from_dob the non zero value
        @param jd_at_dob: Julian day for birthdate and birth time
        @param place: pancganga.Place Struct ('place_name',latitude,longitude,timezone)
        @param dob: Date of birth as a tuple e.g. (1999,12,31)  
        @param years_from_dob: # years of from year of birth
        @param divisional_chart_factor: integer of divisional chart 1=Rasi, 2=D2, 9=D9 etc 
        @return: [lagna_periods,moon_periods,sun_periods]
          Each dhasa period will have the following format:
          [planet index,(dhasa_start_year, month, date,longitude),dhasa duration],...
          [0, (1987, 10, 31, 15.388383474200964), 2.5], [1, (1987, 11, 3, 4.348383475095034), 2.5],....
          
    """
    jd_at_years = jd_at_dob + (years_from_dob * const.sidereal_year)
    rasi_planet_positions = charts.divisional_chart(jd_at_years,place,divisional_chart_factor=divisional_chart_factor)
    lagna_house = rasi_planet_positions[0][1][0]
    moon_house = rasi_planet_positions[2][1][0]
    sun_house = rasi_planet_positions[1][1][0]
    print('lagna house natal chart',lagna_house)
    lagna_sign = (lagna_house+years_from_dob-1) % 12
    moon_sign = (moon_house+years_from_dob-1) % 12
    sun_sign = (sun_house+years_from_dob-1) % 12
    print('lagna/Moon/Sun house on annual natal chart',lagna_sign,moon_sign,sun_sign)
    lagna_periods = _sudharsana_dhasa_calculation(jd_at_years,lagna_sign)
    moon_periods = _sudharsana_dhasa_calculation(jd_at_years,moon_sign)
    sun_periods = _sudharsana_dhasa_calculation(jd_at_years,sun_sign)
    return lagna_periods,moon_periods,sun_periods
def _sudharsana_dhasa_calculation(jd_at_years,dhasa_seed_sign):
    dhasa_periods = []
    dhasa_start = jd_at_years
    dhasa_progression = [(dhasa_seed_sign+h)%12 for h in range(12)]
    dhasa_duration = const.sidereal_year
    antardhasa_duration = round(const.sidereal_year / 12.0,2)
    for sign in dhasa_progression:
        dhasa_end = dhasa_start+dhasa_duration
        antardhasa = [[(sign+h)%12,panchanga.jd_to_gregorian(dhasa_start+h*antardhasa_duration),1.0] for h in range(12)]
        dhasa_end_date = panchanga.jd_to_gregorian(dhasa_end)
        dhasa_periods.append([sign,antardhasa,dhasa_end_date,1.0])
        dhasa_start = dhasa_end
    return dhasa_periods
def sudharsana_pratyantardasas(antardhasa_start,antardhasa_seed_sign):
    dhasa_periods = []
    dhasa_start = antardhasa_start
    dhasa_progression = [(antardhasa_seed_sign+h)%12 for h in range(12)]
    dhasa_duration = round(const.sidereal_year/144.0,2)
    for sign in dhasa_progression:
        dhasa_end = dhasa_start+dhasa_duration
        dhasa_end_date = panchanga.jd_to_gregorian(dhasa_end)
        dhasa_periods.append([sign,dhasa_end_date,2.5])
        dhasa_start = dhasa_end
    return dhasa_periods
    

File: pyjhora-web/src/core/dhasa/sudharsana-chakra.ts
/**
 * Sudharsana Chakra Dhasa System
 * Ported from PyJHora sudharsana_chakra.py
 *
 * Sign-based dhasa system with Lagna, Moon, and Sun progressions.
 * Each mahadasha is one sign for one sidereal year, with 12 antardhasas.
 */

import { SIDEREAL_YEAR, RASI_NAMES_EN } from '../constants';
import type { PlanetPosition } from '../horoscope/charts';
import { planetsInRetrograde, getHousePlanetListFromPositions } from '../horoscope/charts';
import { julianDayToGregorian } from '../utils/julian';

// ============================================================================
// TYPES
// ============================================================================

export interface SudharsanaAntardasha {
  sign: number;
  signName: string;
  startJd: number;
  startDate: string;
  durationDays: number;
}

export interface SudharsanaDashaPeriod {
  sign: number;
  signName: string;
  endJd: number;
  endDate: string;
  durationDays: number;
  antardhasas: SudharsanaAntardasha[];
}

export interface SudharsanaPratyantardasha {
  sign: number;
  signName: string;
  endJd: number;
  endDate: string;
  durationDays: number;
}

export interface SudharsanaChakraChart {
  lagnaChart: Array<[number, string]>;
  moonChart: Array<[number, string]>;
  sunChart: Array<[number, string]>;
  retrogradePlanets: number[];
}

export interface SudharsanaChakraDhasaResult {
  lagnaPeriods: SudharsanaDashaPeriod[];
  moonPeriods: SudharsanaDashaPeriod[];
  sunPeriods: SudharsanaDashaPeriod[];
}

// ============================================================================
// HELPERS
// ============================================================================

function formatJdAsDate(jd: number): string {
  const { date, time } = julianDayToGregorian(jd);
  const pad = (n: number) => Math.abs(n).toString().padStart(2, '0');
  return `${date.year}-${pad(date.month)}-${pad(date.day)} ${pad(time.hour)}:${pad(time.minute)}:${pad(time.second)}`;
}

// ============================================================================
// SUDHARSANA CHAKRA CHART
// ============================================================================

/**
 * Create Sudharsana Chakra charts (Lagna, Moon, Sun rotations).
 * @param positions - D1 (or divisional) planet positions
 * @returns Charts rotated from Lagna, Moon, and Sun houses + retrograde planets
 */
export function sudharsanaChakraChart(
  positions: PlanetPosition[],
): SudharsanaChakraChart {
  const retrograde = planetsInRetrograde(positions);
  const natalChart = getHousePlanetListFromPositions(positions);

  const lagnaHouse = positions[0]!.rasi;
  const moonHouse = positions[2]!.rasi;
  const sunHouse = positions[1]!.rasi;

  const lagnaChart: Array<[number, string]> = [];
  const moonChart: Array<[number, string]> = [];
  const sunChart: Array<[number, string]> = [];

  for (let p = 0; p < 12; p++) {
    const lIdx = (p + lagnaHouse) % 12;
    lagnaChart.push([lIdx, natalChart[lIdx]!]);

    const mIdx = (p + moonHouse) % 12;
    moonChart.push([mIdx, natalChart[mIdx]!]);

    const sIdx = (p + sunHouse) % 12;
    sunChart.push([sIdx, natalChart[sIdx]!]);
  }

  return { lagnaChart, moonChart, sunChart, retrogradePlanets: retrograde };
}

// ============================================================================
// DHASA CALCULATION
// ============================================================================

/**
 * Calculate Sudharsana dhasa periods from a seed sign.
 * 12 periods of 1 sidereal year each, progressing through signs.
 * Each period has 12 antardhasas.
 */
function sudharsanaDhasaCalculation(
  jdStart: number,
  seedSign: number,
): SudharsanaDashaPeriod[] {
  const periods: SudharsanaDashaPeriod[] = [];
  let dhasaStart = jdStart;
  const dhasaDuration = SIDEREAL_YEAR;
  const antardasaDuration = Math.round((SIDEREAL_YEAR / 12.0) * 100) / 100;

  for (let h = 0; h < 12; h++) {
    const sign = (seedSign + h) % 12;
    const dhasaEnd = dhasaStart + dhasaDuration;

    const antardhasas: SudharsanaAntardasha[] = [];
    for (let a = 0; a < 12; a++) {
      const antSign = (sign + a) % 12;
      const antStart = dhasaStart + a * antardasaDuration;
      antardhasas.push({
        sign: antSign,
        signName: RASI_NAMES_EN[antSign] ?? `Sign${antSign}`,
        startJd: antStart,
        startDate: formatJdAsDate(antStart),
        durationDays: antardasaDuration,
      });
    }

    periods.push({
      sign,
      signName: RASI_NAMES_EN[sign] ?? `Sign${sign}`,
      endJd: dhasaEnd,
      endDate: formatJdAsDate(dhasaEnd),
      durationDays: dhasaDuration,
      antardhasas,
    });

    dhasaStart = dhasaEnd;
  }

  return periods;
}

/**
 * Main entry: Calculate Sudharsana Chakra dhasas for Lagna, Moon, and Sun.
 * @param positions - Planet positions (D1 or divisional chart)
 * @param jd - Julian Day to start from
 * @param yearsFromDob - Number of years from birth (for annual charts)
 * @returns Lagna, Moon, and Sun dhasa periods
 */
export function getSudharsanaChakraDhasa(
  positions: PlanetPosition[],
  jd: number,
  yearsFromDob: number = 0,
): SudharsanaChakraDhasaResult {
  const lagnaHouse = positions[0]!.rasi;
  const moonHouse = positions[2]!.rasi;
  const sunHouse = positions[1]!.rasi;

  // Seed sign progresses by yearsFromDob
  const lagnaSign = (lagnaHouse + yearsFromDob - 1 + 12) % 12;
  const moonSign = (moonHouse + yearsFromDob - 1 + 12) % 12;
  const sunSign = (sunHouse + yearsFromDob - 1 + 12) % 12;

  const jdAtYears = jd + yearsFromDob * SIDEREAL_YEAR;

  const lagnaPeriods = sudharsanaDhasaCalculation(jdAtYears, lagnaSign);
  const moonPeriods = sudharsanaDhasaCalculation(jdAtYears, moonSign);
  const sunPeriods = sudharsanaDhasaCalculation(jdAtYears, sunSign);

  return { lagnaPeriods, moonPeriods, sunPeriods };
}

/**
 * Calculate pratyantardasas (sub-sub-periods) for a given antardhasa.
 * @param antardhasaStartJd - Start JD of the antardhasa
 * @param antardhasaSeedSign - Seed sign of the antardhasa
 * @returns 12 pratyantardasas
 */
export function sudharsanaPratyantardasas(
  antardhasaStartJd: number,
  antardhasaSeedSign: number,
): SudharsanaPratyantardasha[] {
  const periods: SudharsanaPratyantardasha[] = [];
  let start = antardhasaStartJd;
  const duration = Math.round((SIDEREAL_YEAR / 144.0) * 100) / 100;

  for (let h = 0; h < 12; h++) {
    const sign = (antardhasaSeedSign + h) % 12;
    const end = start + duration;

    periods.push({
      sign,
      signName: RASI_NAMES_EN[sign] ?? `Sign${sign}`,
      endJd: end,
      endDate: formatJdAsDate(end),
      durationDays: duration,
    });

    start = end;
  }

  return periods;
}
